0024-pre-shared-key-v2/0024-pre-shared-key-v2.md

```diff
@@ -0,0 +1,266 @@
+---
+title: Pre-Shared Key V2 (PSKv2) Transport Protocol
+draft: 1
+---
+# Pre-Shared Key V2 (PSKv2) Transport Protocol
+
+This spec defines version 2 of the Pre-Shared Key Transport Protocol. PSKv2 implements end-to-end data encryption, condition and fulfillment generation, [End-to-End Quoting](#end-to-end-quoting) and [Chunked Payments](#chunked-payments). The approach to chunked payments is inspired by the Internet's Transmission Control Protocol (TCP).
+
+## Overview
+
+Like [PSKv1](./0016-pre-shared-key/0016-pre-shared-key.md), PSKv2 uses a secret shared between the sender and receiver to generate the conditions and fulfillments for payments without repeated end-to-end communication. PSKv2 additionally includes support for End-to-End Quoting and Chunked Payments.
+
+This specificiation aims to standardize the behavior of receivers while allowing for sophisticated sender behavior to be developed in the future. [Appendix A: Advanced Sender Features](#appendix-a-advanced-sender-features) outlines areas for potential sender features to be implemented in the future.
+
+## Setup
+
+Before using PSKv2, the sender and receiver use an authenticated, encrypted communication channel (generally provided as part of an [Application Layer Protocol](./0001-interledger-architecture/0001-interledger-architecture.md#application-layer)) to exchange:
+
+- PSK version (such as `2.0` for this version)
+- A 32-byte random or pseudorandom shared secret
+- The receiver's ILP address
+
+The receiver MAY generate the shared secret from a longer-term secret and a nonce that they put into the ILP address they give to the sender, as described in [PSKv1's Appendix A: Recommended Algorithm for Generating Shared Secrets](./0016-pre-shared-key/0016-pre-shared-key.md#appendix-a-recommended-algorithm-for-generating-shared-secrets).
+
+### Encryption
+
+In PSKv2.0, request and response data is encrypted using [AES-256-GCM](https://en.wikipedia.org/wiki/Galois/Counter_Mode) with a 16-byte Initialization Vector (IV) and a 16-Byte Authentication Tag.
```

Comment:
Regarding the size of the Authentication Tag. This is an excerpt from the [NIST publication](http://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-38d.pdf) on GCM:
> The  bit  length  of  the  tag,  denoted  t,  is  a  security  parameter,  as  discussed  in  Appendix  B. In general, t may  be  any  one  of  the  following  five values:  128,  120,  112,  104,  or  96.    For  certain applications, t  may  be  64  or  32;  guidance  for  the  use  of  these  two  tag  lengths, including requirements on the length of the input data and the lifetime of the key in these cases, is given in Appendix C.

So, a 16 byte Authentication Tag can be considered secure.

Another source recommending a 16 byte authentication tag is _Cryptography Engineering: Design Principles and Practical Applications_ (p. 309):
> “Therefore, although it is possible to reduce the size of the authenticator for GCM from 128 bits to something less, we recommend not doing so. Our recommendation is to only use GCM with the full 128-bit authentication tag.”

Reply:
Ok, and I think the recommendation is also to use a 16-byte IV, right?

Reply:
The NIST publication recommends a bit length of 96 for the IV:
> For  IVs,  it  is  recommended  that  implementations restrict  support  to  the length  of  96  bits,  to promote interoperability, efficiency, and simplicity of design.

Reply:
Note that the [Rust](https://briansmith.org/rustdoc/ring/aead/static.AES_256_GCM.html), [Go](https://golang.org/pkg/crypto/cipher/#AEAD), and [Java](https://stackoverflow.com/questions/31851612/java-aes-gcm-nopadding-what-is-cipher-getiv-giving-me) implementations either require or recommend using a 96-bit IV as well.